DMA (eDMA) and UART+DMA (Async API support) Enablement

Shreyas-Shankar155 opened the pull request:
This PR includes:

-> eDMA driver
-> SoC level Clock and XBAR enablements
-> Modifications to uart_ns16550 to enable DMA support for TI's UART IP

Shreyas-Shankar155 committed:
soc: am261x: Clocks and XBAR enablement

- Add logic to enable UART clocks for every enabled instance in dts.

- Enable DMA XBARs for necessary peripherals.
  Currently, only UART DMA XBAR selection is supported.
  DMA Channel is specified by DTS / appropriate overlay file.

Fixes: ASMZEP-45, ASMZEP-47

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
soc: am261x: Add FPU Symbols

Add FPU symbols in SoC layer.
This is to ensure ABI compatibility with TI's MCU SDK ABIs.

NOTE: VFP_DP_D16 symbol is selected, it signifies
      SoC uses a double-precision VFP processor.

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
drivers: dma: Add eDMA driver

Add driver code, headers and bindings file

This DMA implementation supports M2M / P2M / M2P transfer directions.
HAL is used for eDMA configuration.
Interrupt-registration is handled by Zephyr.

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
soc: am261x: Add DMA support

- Select DMA symbol in Kconfig file
- Mention Number of EDMAs in SoC

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
dts: am261x: Add DMA

Add eDMA in AM261x DTS.

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
boards: am261x_lp: Add DMA configuration

Activate eDMA in boards area with sample configuration.

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
soc: ti: am261x: Enable Serial Async

Enable SERIAL_SUPPORT_ASYNC for AM261x SoC for NS16550 UART driver

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
drivers: serial: ns16550: Add TI K3 DMA support

- Enable DMA mode in UART driver for TI K3 platforms
- Modify Kconfig for Async API support
- Add macros to dedicated TI_K3 header file
- Add conditional compilation for TI K3 DMA functionality

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

natto1784 commented on drivers/dma/Kconfig.ti_edma:
```diff
@@ -0,0 +1,16 @@
+	int "Number of EDMA IP present on SoC"
+	default 1
+	help
+		Number of EDMA IPs present on SoC
```
extra whitespace

natto1784 commented on drivers/serial/uart_ns16550.c:
```diff
@@ -566,6 +588,165 @@ static void set_baud_rate(const struct device *dev, uint32_t baud_rate, uint32_t
+			 UART_IER_SLEEP_MODE_SHIFT);
+
+	UART_regConfigModeEnable(baseAddr, UART_REG_CONFIG_MODE_B);
+	divRegVal;
```
artifact

Shreyas-Shankar155 replied:
removed

natto1784 commented on drivers/serial/uart_ns16550.c:
```diff
@@ -566,6 +588,165 @@ static void set_baud_rate(const struct device *dev, uint32_t baud_rate, uint32_t
 	}
 }
 
+#ifdef CONFIG_UART_NS16550_TI_K3
```
this is a big ifdef, would it be better to separate it out into its own file?

Shreyas-Shankar155 replied:
i'm open to discussion on this.

this ifdef prepares the UART IP for DMA mode. and i believe it's K3 IP specific.

natto1784 commented on include/zephyr/drivers/dma/dma_ti_edma.h:
```diff
@@ -0,0 +1,39 @@
+    struct EDMA_ISR_Data isr_data;
+};
+
+#endif // __TI_EDMA_HEADERS__
```
newline required before EOF

Shreyas-Shankar155 replied:
fixed

natto1784 commented on include/zephyr/drivers/serial/uart_ns16550_ti_k3.h:
```diff
@@ -0,0 +1,114 @@
+/*
```
are these values unique to TI NS16550 IP? If not, then might be a better idea to put them in the main driver file.

natto1784 commented on soc/ti/am26x/am261x/common/clocks_configs.c:
```diff
@@ -0,0 +1,38 @@
+		sys_write32(0x777, CSL_MSS_RCM_LIN5_UART5_CLK_SRC_SEL); 
+	#endif
+
+}
```
newline required before EOF

Shreyas-Shankar155 replied:
fixed

natto1784 commented on soc/ti/am26x/am261x/common/xbar_configs.c:
```diff
@@ -0,0 +1,93 @@
+#ifdef CONFIG_DMA
+void configure_dma_xbars() {
+
+	#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart0), okay)
```
preprocessor directives will leave whitespace behind after translation, please move to the beginning of line. more opinions required on this.

Shreyas-Shankar155 replied:
i ran the clang-format script on these.
it moved the ifdefs to beginning of line.

natto1784 commented on soc/ti/am26x/am261x/common/xbar_configs.c:
```diff
@@ -0,0 +1,93 @@
+
+#include <xbar_configs.h>
+
+#ifdef CONFIG_DMA
```
rather than putting entire file within ifdef, better to source it conditionally in the cmakelists.txt by checking `CONFIG_DMA`

Shreyas-Shankar155 replied:
This is a placeholder for future XBAR configs as well.
Not all XBAR configs are DMA-related

natto1784 replied:
then it would be better for it to have its own kconfig option

Shreyas-Shankar155 replied:
meaning use a new symbol like CONFIG_DMA_XBAR instead of just CONFIG_DMA?

what advantage does that provide over the current implementation?

natto1784 replied:
Just a suggestion for replacing `CONFIG_DMA` which is a placeholder with something more futureproof

natto1784 commented on soc/ti/am26x/am261x/r5/soc.c:
```diff
@@ -65,4 +67,8 @@ void soc_early_init_hook(void)
+	configure_soc_clocks();
+	#ifdef CONFIG_DMA
+		configure_dma_xbars();
+	#endif
```
same as above, unnecessary whitespace

Shreyas-Shankar155 replied:
fixed

Shreyas-Shankar155 commented:
<img width="2055" height="261" alt="fff497f5-a5fc-4eaa-9e66-b1f1d01ab88e" src="https://github.com/user-attachments/assets/ca69828a-bf4f-47fe-9e2b-a89d063658a6" />



the compliance check keeps throwing above error, telling me to run clang-format.
However, the moment i do that, it says the lines exceed 100chars.

Also, the script adds an extra space before the closing parenthesis in first line:
EDMA_RESOURCE_GET, (,[space])

natto1784 commented:
> the compliance check keeps throwing above error, telling me to run clang-format. However, the moment i do that, it says the lines exceed 100chars.

this is something that I also face at times, with such macros, you must format by hand and can't rely on clang-format. one such case `(,)` which is correct but clang-format changes it to `(, )`

Shreyas-Shankar155 committed:
boards: ti: am261x_lp: Remove interrupt-driven Kconfig

Remove UART_INTERRUPT_DRIVEN Kconfig option.
These should be selected by application, NOT board-level Kconfig.

Add supported hardware in board-level yaml files.

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

Shreyas-Shankar155 committed:
dts: am261x: Add dma cells

Add dma cells for UART in dts.

Signed-off-by: Shreyas Shankar <s-shankar@ti.com>

MihiraMadhava commented:
https://github.com/TexasInstruments/asm-hal_ti/pull/1

this is a pre-requisite for this PR merge.

MihiraMadhava committed:
TI: west: asm-hal_ti: update hal_ti to asm-hal_ti

Updating the west.yml to pick asm-hal_ti instead of hal_ti.
added the downstream remote for
https://github.com/TexasInstruments/asm-hal_ti.git and updated the
revision to the latest release tag v4.2.0-ti-11.00.01

Signed-off-by: Mihira Madhava Bollapragada <madhava@ti.com>
